component/file_cache/file_cache_async.go

```diff
@@ -0,0 +1,331 @@
+
+					case fileOperation == "ChmodAndFlush":
+						returnVal = fc.asyncChmodAndFlush(fileOptions.(internal.ChmodOptions))
+					}
+
+					log.Trace("AsyncFileCache:: async_cloud_handler: The key after the function call is %s and the value is %s", key, fileOptions)
+
+					if returnVal == nil {
+						log.Trace("AsyncFileCache:: async_cloud_handler: File name %s has just finished file operation %s", key, fileOperation)
+						tries = 0                                 //attempt was successful, reset try counter
+						_ = fc.fileOps.CompareAndDelete(key, val) //file has been serviced, remove it from map only if file op hasn't been updated
+
+					} else {
+
+						numFailed++
+						if tries < maxTries {
+
+							tries++ //failed op, increase timeout duration
+						}
+
+					}
+				}
+				return true
+			})
+		}
+
+		//check channel to end thread here
+		// if <-fc.closeSignal == 1 {
+		// 	return
+		// }
+	}
+
+}
+
+// File is already flushed locally, we just need to upload it to the cloud
+func (fc *FileCache) asyncFlushFile(options FlushFileAbstraction) error {
+
+	localPath := common.JoinUnixFilepath(fc.tmpPath, options.Name)
+	uploadHandle, err := common.Open(localPath)
+
+	if err != nil {
+		log.Err("FileCache::FlushFile : error [unable to open upload handle] %s [%s]", options.Name, err.Error())
+		return err
+	}
+
+	err = fc.NextComponent().CopyFromFile(
+		internal.CopyFromFileOptions{
+			Name: options.Name,
+			File: uploadHandle,
+		})
+
+	uploadHandle.Close()
+	if err != nil {
+		log.Err("FileCache::FlushFile : %s upload failed [%s]", options.Name, err.Error())
+		return err
+	}
+
+	//use miss list to update mode, take it out of flush file
+
+	return nil
+}
+
+func (fc *FileCache) asyncDeleteFile(options internal.DeleteFileOptions) error {
+
+	err := fc.NextComponent().DeleteFile(options)
+	err = fc.validateStorageError(options.Name, err, "DeleteFile", false)
+	if err != nil {
+		log.Err("FileCache::DeleteFile : error  %s [%s]", options.Name, err.Error())
+		return err
+	}
+	return nil
+}
+
+func (fc *FileCache) asyncRenameFile(options internal.RenameFileOptions) error {
+
+	var getAttrOptions internal.GetAttrOptions
+
+	getAttrOptions.Name = options.Src
+	getAttrOptions.RetrieveMetadata = false //not sure what this is for
+
+	_, err := fc.NextComponent().GetAttr(getAttrOptions)
+
+	if err != nil && err == syscall.ENOENT { //src file does not exist in cloud
+
+		err = fc.asyncFlushFile(FlushFileAbstraction{Name: options.Dst})
+		if err != nil {
+			log.Err("FileCache::RenameFile : %s failed to flush file [%s]", options.Dst, err.Error())
+			return err
+		}
+		return nil
+
+	} else {
+
+		err = fc.NextComponent().RenameFile(options)
+		err = fc.validateStorageError(options.Src, err, "RenameFile", false)
+		if err != nil {
+			log.Err("FileCache::RenameFile : %s failed to rename file [%s]", options.Src, err.Error())
+			return err
+		}
+		return nil
+
+	}
+	// err = fc.NextComponent().RenameFile(options)
```

Comment:
Is this commented code section for debug?

Reply:
It was just a section of code I needed, but in a different place. Removed that comment!